libs/wingc/src/jsify.rs

```diff
@@ -41,21 +42,190 @@ const $App = __app(process.env.WING_TARGET);
+		Self {
+			references: BTreeMap::new(),
+			diagnostics: Diagnostics::new(),
+			method_name,
+			path: vec![],
+			function_def,
+			env: function_def.statements.env.borrow().as_ref().unwrap().get_ref(),
+		}
+	}
+
+	pub fn find_refs(mut self) -> (BTreeMap<String, IndexSet<String>>, Diagnostics) {
+		self.visit_scope(&self.function_def.statements);
+		(self.references, self.diagnostics)
+	}
+}
+
+impl<'ast> Visit<'ast> for FieldReferenceVisitor<'_> {
+	fn visit_reference(&mut self, node: &'ast Reference) {
+		match node {
+    	Reference::InstanceMember { object, property } => {
+				match &object.kind {
+        	ExprKind::Reference(r) => {
+						match r {
+							Reference::Identifier(s) => {
+								if s.name == "this" {
+									// resolve the type of the resource by looking up "this" in our environment
+									let resource_type = self.env
+										.lookup(s, None).expect("'this' to be found")
+										.as_variable().expect("'this' to be a variable")
+										.type_;
+
+									// lookup our field in the class environment
+									match &*resource_type {
+            				Type::Resource(r) => {
+											let field_kind = r.env
+												.lookup(property, None).expect("unable to find field in resource env")
+												.as_variable().expect("field is not a variable");
+
+											// we only care about preflight fields (inflight fields are normal references)
+											if field_kind.flight == Phase::Preflight {
+
+												// don't allow capturing reassignable (`var`) fields.
+												if field_kind.reassignable {
+													self.diagnostics.insert(Diagnostic {
+														level: DiagnosticLevel::Error,
+														message: format!(
+															"Unable to reference \"this.{}\" from inflight method \"{}\" because it is reassignable (\"var\")",
+															property.name, 
+															self.method_name.name,
+														),
+														span: Some(property.span.clone()),
+													});
+													return;
+												}
+
+												// check if the type is capturable (resource, primitive or immutable collection)
+												if ! field_kind.type_.is_capturable() {
+													self.diagnostics.insert(Diagnostic {
+														level: DiagnosticLevel::Error,
+														message: format!(
+															"Unable to reference \"this.{}\" from inflight method \"{}\" because type {} is not capturable",
+															property.name, 
+															self.method_name.name,
+															field_kind.type_,
+														),
+														span: Some(property.span.clone()),
+													});
+
+													return;
+												}
+
+												debug!("field \"this.{}\" is referenced from inflight method {}", property.name, self.method_name.name);
+
+												// get/create an "ops" set for this field
+												let mut ops = if let Some(ops) = self.references.get(&property.name) {
+													ops.clone()
+												} else {
+													IndexSet::new()
+												};
```

Comment:
The idiomatic way to get-and-mutate a value in a map in Rust without having to call `get` and `insert` two separate times is to use the [Entry](https://doc.rust-lang.org/stable/std/collections/hash_map/enum.Entry.html) API -- I think it can probably be used here? 
```suggestion
												let mut ops = self.references.entry(property.name.clone()).or_default();
```

Reply:
👍